service/tasks/monitoring.py

```diff
@@ -118,8 +121,41 @@ def prune_machines_for(provider_id, print_logs=False, dry_run=False, forced_remo
     for machine in db_machines:
         cloud_match = [mach for mach in cloud_machine_ids if mach == machine.identifier]
         if not cloud_match:
-            remove_machine(machine, dry_run=dry_run)
-
+            remove_machine(machine, now, dry_run=dry_run)
+            mach_count += 1
+    # Loop 2 and 3 - Capture all (still-active) versions without machines, and all applications without versions.
+    # These are 'outliers' and mainly here for safety-check purposes.
+    versions_without_machines = ApplicationVersion.objects.filter(machines__isnull=True, end_date__isnull=True)
+    apps_without_versions = Application.objects.filter(versions__isnull=True, end_date__isnull=True)
+    for ver in versions_without_machines:
+        ver.end_date_all(now)
+        ver_count += 1
+    for app in apps_without_versions:
+        app.end_date_all(now)
+        app_count += 1
+
+    #Loop4 - All 'Application' DB objects require >=1 Version with >=1 ProviderMachine (ACTIVE!)
+    # Apps that don't meet this criteria should be end-dated.
+    improperly_enddated_apps = Application.objects.annotate(
+        num_versions=Count('versions'), num_machines=Count('versions__machines')
+    ).filter(
+        # Contains at least one version without an end-date OR
+        (Q(num_versions__gt=0) & Q(versions__end_date__isnull=True)) |
+        # conatins at least one machine without an end-date
+        (
+            Q(num_machines__gt=0) &
+            Q(versions__machines__instance_source__end_date__isnull=True)
+        ),
+        # AND application has already been end-dated.
+        end_date__isnull=False
+    )
```

Comment:
If worth extracting this query into a function? It might not be. But `prune_machine_for` has varying levels of operational abstraction ... or could be abstracted (might be a better way to put it).